### PR TITLE
Make TSX Work Once Again

### PR DESCRIPTION
**Pull request by Nekotekina**
~~Unfinished, test only with TSX enabled.~~

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellSpurs.cpp [hunk 1]
```diff
@@ -4706,24 +4709,28 @@ s32 cellSpursJobGuardNotify(ppu_thread& ppu, vm::ptr<CellSpursJobGuard> jobGuard
...
 	u32 allow_jobchain_run = 0; // Affects cellSpursJobChainRun execution
+	u32 old = 0;
 
-	auto [old, ok] = jobGuard->ncount0.fetch_op([&](be_t<u32>& value)
+	const bool ok = vm::reservation_light_op(jobGuard->ncount0, [&](atomic_be_t<u32>& v)
 	{
-		allow_jobchain_run = jobGuard->zero;
-
-		if (!value)
+		// TODO: does it fit for light op?
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellSpurs.cpp:4717]: Nope it doesnt fit, jobGuard->zero is being read you are right.

File: rpcs3/Emu/Cell/Modules/cellSpurs.cpp [hunk 2]
```diff
@@ -4759,9 +4766,12 @@ s32 cellSpursJobGuardReset(vm::ptr<CellSpursJobGuard> jobGuard)
 	if (!jobGuard.aligned())
 		return CELL_SPURS_JOB_ERROR_ALIGN;
 
-	auto [res, rtime] = vm::reservation_lock(jobGuard.addr(), 128, vm::dma_lockb);
-	jobGuard->ncount0 = jobGuard->ncount1;
-	res.release(rtime + 128);
+	vm::reservation_light_op(jobGuard->ncount0, [&](atomic_be_t<u32>& ncount0)
+	{
+		// TODO: is it correct for light op?
+		ncount0 = jobGuard->ncount1;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellSpurs.cpp:4772]: ncount1 is supposed to be some kind of constant value to save the original count's value, realhw only uses stw so its safe to use vm::reservation_light_op here.

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 3]
```diff
@@ -1654,24 +1682,53 @@ static bool ppu_store_reservation(ppu_thread& ppu, u32 addr, u64 reg_value)
...
+			{
+				return false;
+			}
+
+			if (_ok)
+			{
+				break;
+			}
+
+			busy_wait(100);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUThread.cpp:1721]: Now that its not a single try to lock but a loop cpu_thread::check_state() must be called here. Previously it was doing vm::passive_unlock/lock for it but check_state is more optimized.

**Comment by Nekotekina** [comment]
Fixed non-tsx path.

**Comment by digitaldude555** [comment]
Master Screenshot (80)
PRScreenshot (83)
Just to show changes with tsx fa, fps is back.

**Pull request merged**